Add cassandra TLS support

aoby opened the pull request:
* Consolidate duplicate Cassandra setup and config into a new common/cassandra package
* Add TLS support for Cassandra
* Add TLS flags to cadence-cassandra-tool
* Combine TLS config structs into new common auth.TLS
* Add more TLS options:
   - CaFile
   - EnableHostVerification
   - Remove TLS bundleFile option in favor of caFile option

solve #2681

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/cadence?pullRequest=2769) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/cadence?pullRequest=2769) before we can accept your contribution.<br/><hr/>**Andrew Oberhardt** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/cadence?pullRequest=2769) it.</sub>

wxing1292 commented on common/persistence/cassandra/cassandraHelpers.go:
```diff
@@ -92,7 +65,7 @@ func DropCassandraKeyspace(s *gocql.Session, keyspace string) (err error) {
 func LoadCassandraSchema(
-	dir string, fileNames []string, hosts []string, port int, keyspace string, override bool,
+	dir string, fileNames []string, hosts []string, port int, keyspace string, override bool, useTLS bool, tlsCertPath, tlsKeyPath string,
 ) (err error) {
```
plz format like this:
```go
func LoadCassandraSchema(
	dir string, 
	fileNames []string, 
	hosts []string, 
	port int, 
	keyspace string, 
	override bool, 
	useTLS bool, 
	tlsCertPath string, 
	tlsKeyPath string,
) error {
```

wxing1292 replied:
it will be easier to read

wxing1292 commented on common/persistence/cassandra/cassandraHistoryPersistence.go:
```diff
@@ -25,6 +25,8 @@ import (
+
 	"github.com/gocql/gocql"
 
 	workflow "github.com/uber/cadence/.gen/go/shared"
```
plz merge from master then do 
```
make fmt
```
it will sort all imports correctly

wxing1292 commented on common/service/config/config.go:
```diff
@@ -188,6 +188,16 @@ type (
+		CertPath string `yaml:"certPath"`
+		// Optional path to the key file for a TLS client certificate
+		KeyPath string `yaml:"keyPath"`
 	}
```
can you put the TLS config in 'common/auth'
and reuse this definition: https://github.com/uber/cadence/blob/master/common/messaging/kafkaConfig.go#L38
?

wxing1292 commented on tools/cassandra/cqlclient.go:
```diff
@@ -47,6 +49,9 @@ type (
 		numReplicas int
+		UseTLS      bool
+		TLSCertPath string
+		TLSKeyPath  string
```
pls use the above TLS config for simplicity, like
```go
CQLClientConfig struct {
		Hosts       string
		Port        int
		User        string
		Password    string
		Keyspace    string
		Timeout     int
		numReplicas int
		TLS *auth.TLS
}
```

wxing1292 commented on tools/cli/admin.go:
```diff
@@ -64,6 +64,18 @@ func newAdminWorkflowCommands() []cli.Command {
 					Usage: "cassandra keyspace",
 				},
+				cli.BoolFlag{
+					Name:  FlagUseTLS,
```
can we change this to `FlagEnableTLS` ?

wxing1292 commented:
the rest LGTM

meiliang86 commented on common/cassandra/cassandraCluster.go:
```diff
@@ -0,0 +1,77 @@
+	"strings"
+
+	"github.com/gocql/gocql"
+
```
Remove extra empty line.

meiliang86 commented on common/persistence/cassandra/cassandraHelpers.go:
```diff
@@ -92,7 +65,7 @@ func DropCassandraKeyspace(s *gocql.Session, keyspace string) (err error) {
 // LoadCassandraSchema loads the schema from the given .cql files on this keyspace
 func LoadCassandraSchema(
-	dir string, fileNames []string, hosts []string, port int, keyspace string, override bool,
+	dir string, fileNames []string, hosts []string, port int, keyspace string, override bool, useTLS bool, tlsCertPath, tlsKeyPath string,
```
Seems like this method is only used for testing. Can you make it private?

aoby commented:
Thanks @wxing1292 and @meiliang86 for the quick feedback!
I made the requested changes.

I realized in local testing that I needed the settings in `cadence-cassandra-tool` for schema setup, so I added CLI options there as well. I also added settings for the remaining Cassandra tls/ssl [options](https://godoc.org/github.com/gocql/gocql#SslOptions), `CaPath` and `EnableHostVerification`.

Please take another look

wxing1292 commented on common/auth/tls.go:
```diff
@@ -0,0 +1,42 @@
+
+		BundleFile string `yaml:"bundleFile"`
+
+		CaFile string `yaml:caFile` //optional depending on server config
```
these 2 variables are the same. let us keep "CaFile string `yaml:caFile`" for readability.
but before landing this, let me check with one customer who is already using TLS with kafka

wxing1292 replied:
plz also do a `make fmt` locally

aoby replied:
I ran `make fmt` earlier, and it made some formatting changes (included in the last commit).
I just ran it again with no change. Is there something you expect to be reformatted?

wxing1292 replied:
one more thing, the `Contributor License Agreement` does not really recognize your name. one thing you could try is to squash all commits and force push

wxing1292 replied:
> I ran `make fmt` earlier, and it made some formatting changes (included in the last commit).
> I just ran it again with no change. Is there something you expect to be reformatted?

if that is the case, then plz just ignore

aoby replied:
> one more thing, the `Contributor License Agreement` does not really recognize your name. one thing you could try is to squash all commits and force push

I think because I made the first commit from my work account, which is private. I tried to re-author it but that apparently didn't work. I'll try the squash & force push...

wxing1292 commented:
there are also 4 files to be changed:
https://github.com/uber/cadence/blob/master/config/development_active.yaml#L101
https://github.com/uber/cadence/blob/master/config/development_other.yaml#L101
https://github.com/uber/cadence/blob/master/config/development_standby.yaml#L101
https://github.com/uber/cadence/blob/master/tools/cli/admin.go#L376

vancexu commented on common/messaging/kafkaConfig.go:
```diff
@@ -22,26 +22,20 @@ package messaging
-		Enabled    bool   `yaml:"enabled"`
-		CertFile   string `yaml:"certFile"`
-		KeyFile    string `yaml:"keyFile"`
-		BundleFile string `yaml:"bundleFile"`
```
This will be a breaking change for users who use TLS in Kafka. 
@wxing1292 Do you know if any customer is using this? Can we notice them and get this change in?

Otherwise for safety, I would keep it for kafka, and use the new auth.TLS for Cassandra and everything else. 
Or just use the old name "bundleFile"

wxing1292 replied:
talked to our customer, they are ok
